File: L10/knightjump.py
from collections import deque

jump_dir = {(-2, 1), (-2, -1), (1, 2), (1, -2), (-1, 2), (2, 1), (2, -1), (-1, -2)}
n = int(input())
arr = [[c for c in input()] for _ in range(n)]
visited = [[False for __ in range(n)] for _ in range(n)]

def valid_pos(x, y):
    if x < 0 or x >= n:
        return False
    if y < 0 or y >= n:
        return False
    if arr[x][y] != '.':
        return False
    if visited[x][y]:
        return False
    return True
    

# Find knight
x, y = (-1, -1)
for i in range(n):
    for j in range(n):
        if arr[i][j] == 'K':
            x, y = i, j
            break
        
# Queue possible moves
q = deque()
q.append((x, y, 0))
while q:
    x, y, score = q.popleft()
    if x == 0 and y == 0:
        print(score)
        exit()
    for dir in jump_dir:
        r, c = dir
        if valid_pos(r + x, c + y):
            q.append((r+x, c+y, score+1))
            visited[x][y] = True
            
print(-1)

File: L10/knightjump.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <tuple>

using namespace std;

int main()
{
    pair<int,int> jump_dir[] = { make_pair(-2, 1),
                        make_pair(-2, -1),
                        make_pair(1, 2), 
                        make_pair(1, -2), 
                        make_pair(2, 1),
                        make_pair(2, -1),
                        make_pair(-1, -2),
                        };
    int n;
    cin >> n;
    int x, y;
    vector<string> arr;
    vector<vector<bool>> visited;
    string tmp;
    for (int i = 0; i < n; i++)
    {
        cin >> tmp;
        arr.push_back(tmp);
        size_t found = tmp.find('K');
        if (found != string::npos)
        {
            x = i;
            y = found;
        }
        vector<bool> tmp_vect;
        for (int j = 0; j < n; j++)
            tmp_vect.push_back(false);
        visited.push_back(tmp_vect);
    }

    deque<tuple<int, int, int>> q;
    q.push_back(make_tuple(x, y, 0));
    while (q.size() > 0)
    {
        x = get<0>(q.front());
        y = get<1>(q.front());
        int score = get<2>(q.front());
        if (x == 0 && y == 0)
        {
            cout << score;
            return 0;
        }
        for (auto p : jump_dir)
        {
            int tmp_x = p.first + x;
            int tmp_y = p.second + y;
            if (tmp_x >= 0 && tmp_x < n && tmp_y >= 0 && tmp_y < n && arr[tmp_x][tmp_y] == '.' && !visited[tmp_x][tmp_y])
            {
                q.push_back(make_tuple(tmp_x, tmp_y, score + 1));
                visited[x][y] = true;
            }
        }
        q.pop_front();
    }
    cout << -1;
    return 0;
}
